url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/math/Histogram.java

```diff
@@ -0,0 +1,373 @@
+     * @param quanta Number of quanta (bins).
+     */
+    public Histogram(final double lbd, final double ubd, final int quanta) {
+        //
+        this.lbd     = Math.min(lbd,  ubd);
+        this.ubd     = Math.max(lbd,  ubd);
+        this.range   = this.ubd - this.lbd;
+        this.irange  = (this.range > 0.0)?(1.0 / this.range):(0.0);
+        //
+        this.quanta  = Math.max(1, quanta);
+        this.dquanta = ((double)this.quanta) - Double.MIN_VALUE;
+        this.accu    = new double[this.quanta];
+    }
+    
+    /**
+     * Check, if at least one of the histogram values is positive (> 0.0).
+     * <br></br>
+     * @return True if one value is positive, false otherwise.
+     */
+    public boolean hasPositiveValues() {
+        for (int i = 0; i < this.quanta; i++) {
+            if (this.accu[i] > 0.0) return true;
+        }
+        return false;
+    }
+    
+    /**
+     * Check, if at least one of the histogram values is negative (< 0.0).
+     * <br></br>
+     * @return True if one value is negative, false otherwise.
+     */
+    public boolean hasNegativeValues() {
+        for (int i = 0; i < this.quanta; i++) {
+            if (this.accu[i] < 0.0) return true;
+        }
+        return false;
+    }
+    
+    /**
+     * This method produces a simple bar-chart representation of the
+     * inner accumulator as a white-space aligned String. The given
+     * height defines the number of text lines of which the highest bar
+     * in the histogram may consists of. Note, that the resulting
+     * string is just for ascii-style printouts, it should not be
+     * used for extracting histogram data.
+     * <br></br>
+     * @param height The number of lines of the highest bar.
+     * @return A string containing the bar-char representation of the histogram. 
+     */
+    public String printBars(final int height) {
+        //
+        StringBuilder out = new StringBuilder();
+        
+        final boolean neg = this.hasNegativeValues(); 
+        final boolean pos = this.hasPositiveValues(); 
+        //
+        if (pos) {
+            String[][] matrix = new String[height][this.quanta];
+            //
+            double max = 0.0;
+            for (int i = 0; i < this.quanta; i++) {
+                if (accu[i] > max) { 
+                    max = accu[i];
+                }
+            }
+            //
+            final double step = (max / ((double)height));
+            //
+            for (int i = 0; i < this.quanta; i++) {
+                
+                double thres = 0;
+                double value = this.accu[i];
+                
+                for (int j = 0; j < height; j++) {
+                    if (value > thres) {
+                        matrix[j][i] = " || ";
+                    } else {
+                        matrix[j][i] = "    ";
+                    }
+                    thres += step;
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 297:310, 337:350.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.